src/benchmarks/src/generic_bench.cpp

```diff
@@ -0,0 +1,251 @@
+    std::this_thread::sleep_for(std::chrono::microseconds(delay_micros));
+  }
+}
+
+void report_and_clear_metrics(
+    std::unordered_map<std::string, std::string> &config) {
+  int report_delay_seconds = get_int(REPORT_DELAY_SECONDS, config);
+  std::string latency_obj_string = get_str(LATENCY_OBJECTIVE, config);
+  std::string batch_delay_string = get_str(REQUEST_BATCH_DELAY_MICROS, config);
+  std::string reports_path = get_str(REPORTS_PATH, config);
+
+  // Write out run details
+  std::ofstream reports(reports_path);
+  std::stringstream ss;
+  ss << "---Configuration---" << std::endl;
+  for (auto it : config) ss << it.first << ": " << it.second << std::endl;
+  ss << "-------------------" << std::endl;
+  std::string configuration_details = ss.str();
+  reports << configuration_details;
+  log_info("BENCH", configuration_details);
+
+  std::string window, metrics;
+  int window_num = 1;
+
+  // This will get interrupted
+  while (true) {
+    // Wait for metrics
+    std::this_thread::sleep_for(std::chrono::seconds(report_delay_seconds));
+
+    metrics = metrics::MetricsRegistry::get_metrics().report_metrics(true);
+    window = _get_window_str(report_delay_seconds, window_num);
+
+    reports << window << ": " << metrics;
+    reports.flush();
+
+    log_info("WINDOW", window);
+    log_info("METRICS", metrics);
+    window_num += 1;
+  }
+}
+
+void run_benchmark(std::unordered_map<std::string, std::string> &config) {
+  get_config().ready();
+  QueryProcessor qp;
+  std::this_thread::sleep_for(std::chrono::seconds(3));
+
+  clipper::DefaultOutputSelectionPolicy p;
+  clipper::Output parsed_default_output(DEFAULT_OUTPUT, {});
+  auto init_state = p.init_state(parsed_default_output);
+  clipper::StateKey state_key{TEST_APPLICATION_LABEL, clipper::DEFAULT_USER_ID,
+                              0};
+  qp.get_state_table()->put(state_key, p.serialize(init_state));
+
+  // We only need the datapoints – not their labels.
+  std::string cifar_data_path = get_str(CIFAR_DATA_PATH, config);
+  std::unordered_map<int, std::vector<std::vector<double>>> cifar_data =
+      load_cifar(cifar_data_path);
+  auto concatenated_datapoints = concatenate_cifar_datapoints(cifar_data);
+  auto datapoints = concatenated_datapoints.first;
+  auto labels = concatenated_datapoints.second;
+
+  int num_threads = get_int(NUM_THREADS, config);
+  size_t num_datapoints = datapoints.size();
+  unsigned int seed = 0;
+  std::vector<std::vector<int>> indices_for_threads(
+      static_cast<size_t>(num_threads));
+
+  // Assign shuffled datapoint selection indices for each thread
+  for (int j = 0; j < num_threads; j++) {
+    std::vector<int> indices(num_datapoints);
+    for (int i = 0; i < static_cast<int>(num_datapoints); i++) {
+      indices[i] = i;
+    }
+    shuffle(indices.begin(), indices.end(), std::default_random_engine(seed));
+    indices_for_threads.push_back(std::move(indices));
+  }
```

Comment:
`indices[i]` contains an array of integers (running from `0` to `num_datapoints - `) shuffled in a random order. They specify the order in which thread `i` cycles through the data set to send queries.

I thought that each thread should query CIFAR datapoints in a different order (in case the model being tested ran, for some reason, more or less efficiently on a specific datapoint given the ones it processed recently). I'm realizing now that that's not only a strange assumption, but that we also make unique per thread so maybe it's totally irrelevant.

So, we could:
1. Not shuffle at all
2. Since we'll now make copies of the data set for each thread, shuffle the datapoints themselves instead of indices before passing them to each thread.

Do you have a preference? I'm leaning towards the former since I'm now having a hard time thinking of reasons in favor of shuffling.